File: modules/wb_worldgen/ui/lib/mapspace.js
// mapspace.js — shared normalizer for wb_worldgen world map data.
//
// The backend's "world_format 2" replaces the legacy `map` / `map_layers` /
// `map_connections` keys with a flat map space:
//   world_data.maps        object keyed by map_id -> MapRecord
//   world_data.connections flat array of cross-map connection records
//   world_data.root_map_id id of the root map ("root")
// Old API responses (e.g. the map_generation step preview during world
// building) may still carry the legacy shape, so every consumer goes through
// normalizeWorldData() and renders from the normalized form.

/**
 * Normalize any world/map payload (v2 or legacy) to
 * `{ mapsById, rootMapId, connections }`.
 *
 * - v2 (`maps` key): passed through unchanged.
 * - legacy layered (`map_layers`, or step-data `layers` entries carrying a
 *   per-layer `map`): each layer becomes a MapRecord keyed by its layer_id
 *   (lossless — `legacy_layer_id` is preserved), and `map_connections` are
 *   converted to v2 connection records.
 * - legacy flat (`map`): a single "root" MapRecord.
 */
export function normalizeWorldData(worldData) {
  if (!worldData) return { mapsById: {}, rootMapId: null, connections: [] };

  // v2: pass through.
  if (worldData.maps) {
    return {
      mapsById: worldData.maps,
      rootMapId: worldData.root_map_id || 'root',
      connections: worldData.connections || [],
    };
  }

  // Legacy layered: world_data.map_layers, or step-data `layers` whose
  // entries carry a per-layer `map` (same record shape).
  const layers = worldData.map_layers
    || (Array.isArray(worldData.layers) && worldData.layers.some((l) => l?.map)
      ? worldData.layers
      : null);
  if (layers && layers.length) {
    const mapsById = {};
    layers.forEach((layer) => {
      mapsById[layer.layer_id] = {
        map_id: layer.layer_id,
        label: layer.name,
        level_type: layer.layer_type,
        description: layer.description,
        parent_map_id: null,
        anchor_node_id: null,
        legacy_layer_id: layer.layer_id,
        ...(layer.map || {}),
      };
    });
    const rawConns = worldData.map_connections || worldData.connections || [];
    const connections = rawConns.map((c) => {
      // Already v2-shaped (endpoints as {map_id, node_id})? Pass through.
      if (c.from && typeof c.from === 'object') return c;
      return {
        id: c.id,
        from: { map_id: c.from_layer_id, node_id: c.from_node_id },
        to: { map_id: c.to_layer_id, node_id: c.to_node_id },
        kind: c.connection_type,
        name: c.name,
        description: c.description,
        travel: { mode: 'instant' },
        bidirectional: c.bidirectional !== false,
        requirements: c.requirements || [],
        hidden: !!c.hidden,
        origin: c.origin,
      };
    });
    return { mapsById, rootMapId: layers[0].layer_id, connections };
  }

  // Legacy flat map.
  if (worldData.map) {
    return {
      mapsById: {
        root: {
          map_id: 'root',
          label: 'World',
          parent_map_id: null,
          anchor_node_id: null,
          ...worldData.map,
        },
      },
      rootMapId: 'root',
      connections: [],
    };
  }

  return { mapsById: {}, rootMapId: null, connections: [] };
}

/**
 * Index v2 connections by endpoint node id. Returns an object mapping
 * node_id -> array of `{connection, near, far}` views, where `near` is the
 * endpoint at that node and `far` the other end. Bidirectional connections
 * produce a view from each end.
 */
export function connectionsByNode(connections) {
  const byNode = {};
  const add = (nodeId, view) => {
    if (!nodeId) return;
    if (!byNode[nodeId]) byNode[nodeId] = [];
    byNode[nodeId].push(view);
  };
  (connections || []).forEach((connection) => {
    const { from, to } = connection;
    if (!from || !to) return;
    add(from.node_id, { connection, near: from, far: to });
    if (connection.bidirectional !== false) {
      add(to.node_id, { connection, near: to, far: from });
    }
  });
  return byNode;
}

/**
 * Walk parent_map_id links from mapId up to the root and return the chain of
 * MapRecords ordered root -> mapId. Cycle-safe; unknown ids yield [].
 */
export function breadcrumb(mapsById, mapId) {
  const chain = [];
  const seen = new Set();
  let current = mapId;
  while (current && mapsById?.[current] && !seen.has(current)) {
    seen.add(current);
    chain.unshift(mapsById[current]);
    current = mapsById[current].parent_map_id;
  }
  return chain;
}

/**
 * Group child maps by their anchor: `${parentMapId}:${anchorNodeId}` -> [mapId].
 * Maps without a parent or anchor (root / parallel siblings) are excluded.
 */
export function childrenByAnchor(mapsById) {
  const out = {};
  Object.values(mapsById || {}).forEach((m) => {
    if (!m.parent_map_id || !m.anchor_node_id) return;
    const key = `${m.parent_map_id}:${m.anchor_node_id}`;
    if (!out[key]) out[key] = [];
    out[key].push(m.map_id);
  });
  return out;
}

File: modules/wb_worldgen/wbworldgen/worldgen/mapspace.py
"""Accessors over the hierarchical map space (world_format 2).

A v2 world dict carries ``maps`` ({map_id: MapRecord}) and ``connections``
(flat list of ConnectionRecords). These helpers are the single place that
knows that layout; they also tolerate legacy dicts (flat ``map`` or
``map_layers``) so tests and not-yet-migrated inputs keep working.

MapRecord: {map_id, label, level_type, description, parent_map_id,
            anchor_node_id, generator_id, nodes, edges, regions, roads,
            config, legacy_layer_id?, rules?, landmarks?, factions?}
ConnectionRecord: {id, from: {map_id, node_id}, to: {map_id, node_id},
            kind, name, description, travel: {mode, turns?}, bidirectional,
            requirements, hidden, origin}
"""

ROOT_MAP_ID = "root"


def maps_by_id(wd: dict) -> dict:
    """{map_id: MapRecord} view of any world dict (v2 native, legacy adapted).

    Legacy adaptation shares the underlying node/edge lists (no copies), so
    in-place node mutation stays visible through either view.
    """
    if isinstance(wd.get("maps"), dict):
        return wd["maps"]
    result = {}
    map_layers = wd.get("map_layers", [])
    if map_layers:
        for layer in map_layers:
            lid = layer.get("layer_id") or ROOT_MAP_ID
            m = dict(layer.get("map", {}))
            m["map_id"] = lid
            m["label"] = layer.get("name", lid)
            result[lid] = m
        return result
    flat = wd.get("map")
    if isinstance(flat, dict) and flat.get("nodes") is not None:
        m = dict(flat)
        m["map_id"] = ROOT_MAP_ID
        m["label"] = "World"
        return {ROOT_MAP_ID: m}
    return result


def get_map(wd: dict, map_id: str) -> dict | None:
    return maps_by_id(wd).get(map_id)


def all_nodes(wd: dict) -> list[dict]:
    nodes: list[dict] = []
    for m in maps_by_id(wd).values():
        nodes.extend(m.get("nodes", []))
    return nodes


def all_edges(wd: dict) -> list[dict]:
    edges: list[dict] = []
    for m in maps_by_id(wd).values():
        edges.extend(m.get("edges", []))
    return edges


def map_nodes(wd: dict, map_id: str) -> list[dict]:
    m = get_map(wd, map_id)
    return m.get("nodes", []) if m else []


def map_edges(wd: dict, map_id: str) -> list[dict]:
    m = get_map(wd, map_id)
    return m.get("edges", []) if m else []


def node_index(wd: dict) -> dict:
    """{node_id: node} across all maps (node ids are globally unique)."""
    return {n.get("id"): n for n in all_nodes(wd)}


def map_of_node(wd: dict, node_id: str) -> str | None:
    """map_id containing the node, or None."""
    for mid, m in maps_by_id(wd).items():
        for n in m.get("nodes", []):
            if n.get("id") == node_id:
                return mid
    return None


def connections(wd: dict) -> list[dict]:
    return wd.get("connections") or []


def connections_from(wd: dict, map_id: str, node_id: str = None,
                     include_hidden: bool = False) -> list[dict]:
    """Outgoing connection views for a map (optionally one node).

    Bidirectional connections whose far side is on this map are returned
    reversed, so callers always see {connection, near: {map_id, node_id},
    far: {map_id, node_id}} with ``near`` on the requested map.
    """
    out = []
    for c in connections(wd):
        if c.get("hidden") and not include_hidden:
            continue
        frm, to = c.get("from") or {}, c.get("to") or {}
        if frm.get("map_id") == map_id and (node_id is None or frm.get("node_id") == node_id):
            out.append({"connection": c, "near": frm, "far": to})
        elif c.get("bidirectional", True) and to.get("map_id") == map_id \
                and (node_id is None or to.get("node_id") == node_id):
            out.append({"connection": c, "near": to, "far": frm})
    return out


def find_connection(wd: dict, connection_id: str) -> dict | None:
    for c in connections(wd):
        if c.get("id") == connection_id:
            return c
    return None


def connection_between(wd: dict, a: dict, b: dict,
                       include_hidden: bool = True) -> dict | None:
    """A connection directly joining endpoints a/b ({map_id, node_id}), if any."""
    def _same(x, y):
        return x.get("map_id") == y.get("map_id") and x.get("node_id") == y.get("node_id")
    for c in connections(wd):
        frm, to = c.get("from") or {}, c.get("to") or {}
        if c.get("hidden") and not include_hidden:
            continue
        if (_same(frm, a) and _same(to, b)) or (_same(frm, b) and _same(to, a)):
            return c
    return None


def breadcrumb(wd: dict, map_id: str) -> list[dict]:
    """MapRecords from the root down to (and including) map_id."""
    by_id = maps_by_id(wd)
    trail = []
    seen = set()
    cur = by_id.get(map_id)
    while cur is not None and cur.get("map_id") not in seen:
        seen.add(cur.get("map_id"))
        trail.append(cur)
        parent = cur.get("parent_map_id")
        cur = by_id.get(parent) if parent else None
    trail.reverse()
    return trail


def children_by_anchor(wd: dict) -> dict:
    """{(parent_map_id, anchor_node_id): [map_id, ...]} derived index."""
    idx: dict[tuple, list] = {}
    for mid, m in maps_by_id(wd).items():
        parent, anchor = m.get("parent_map_id"), m.get("anchor_node_id")
        if parent and anchor:
            idx.setdefault((parent, anchor), []).append(mid)
    return idx


def connection_endpoints(wd: dict) -> set:
    """{(map_id, node_id)} across every connection's both ends."""
    endpoints = set()
    for c in connections(wd):
        for end in (c.get("from") or {}, c.get("to") or {}):
            if end.get("map_id") and end.get("node_id"):
                endpoints.add((end["map_id"], end["node_id"]))
    return endpoints


def connected_components(nodes: list, edges: list) -> list:
    """Connected components (lists of node ids) over one map's undirected
    edge list, largest first. Edge endpoints not in ``nodes`` are ignored."""
    ids = [n.get("id") for n in nodes if n.get("id")]
    idset = set(ids)
    adjacency = {nid: [] for nid in ids}
    for e in edges:
        a, b = e.get("from"), e.get("to")
        if a in idset and b in idset:
            adjacency[a].append(b)
            adjacency[b].append(a)
    seen = set()
    components = []
    for nid in ids:
        if nid in seen:
            continue
        stack, comp = [nid], []
        seen.add(nid)
        while stack:
            cur = stack.pop()
            comp.append(cur)
            for nb in adjacency[cur]:
                if nb not in seen:
                    seen.add(nb)
                    stack.append(nb)
        components.append(comp)
    components.sort(key=len, reverse=True)
    return components


def unreachable_maps(wd: dict, maps: dict = None) -> set:
    """Map ids not reachable from the root map over connections (either
    direction) and parent-child anchoring. The root is the ``root`` map when
    present, else the first parentless map, else the first map."""
    if maps is None:
        maps = maps_by_id(wd)
    if len(maps) < 2:
        return set()
    start = ROOT_MAP_ID if ROOT_MAP_ID in maps else next(
        (mid for mid, m in maps.items() if not m.get("parent_map_id")),
        next(iter(maps)))
    adjacency: dict = {mid: set() for mid in maps}
    for c in connections(wd):
        a = (c.get("from") or {}).get("map_id")
        b = (c.get("to") or {}).get("map_id")
        if a in adjacency and b in adjacency:
            adjacency[a].add(b)
            adjacency[b].add(a)
    for mid, m in maps.items():
        parent = m.get("parent_map_id")
        if parent in adjacency and m.get("anchor_node_id"):
            adjacency[parent].add(mid)
            adjacency[mid].add(parent)
    seen = {start}
    stack = [start]
    while stack:
        for nb in adjacency[stack.pop()]:
            if nb not in seen:
                seen.add(nb)
                stack.append(nb)
    return set(maps) - seen


def parallel_siblings(wd: dict, map_id: str) -> list[dict]:
    """Other maps sharing this map's parent with no anchor (parallel planes),
    plus the parent-anchored view for the root's parallels."""
    by_id = maps_by_id(wd)
    me = by_id.get(map_id)
    if me is None:
        return []
    parent = me.get("parent_map_id")
    return [m for mid, m in by_id.items()
            if mid != map_id and m.get("parent_map_id") == parent
            and not m.get("anchor_node_id") and not me.get("anchor_node_id")]
